iwyu_ast_util.cc

```diff
@@ -587,18 +593,41 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
     // TODO(bolshakov): it doesn't always work properly. For example, both A and
     // TypedefForA would be inserted in seen_types_ for Tpl<A, TypedefForA>,
     // leading to indeterminate resugar_map construction.
-    // TODO(bolshakov): check whether typedef components are "provided".
-    return Base::TraverseType(QualType(desugared, 0));
+    return Base::Base::TraverseType(QualType(desugared, 0));
   }
 
-  set<const Type*> seen_types_;
+ private:
+  typedef TypeEnumerator<TemplateArgEnumerator> Base;
+};
+
+class TypeEnumeratorWithoutSubstituted
```

Comment:
These two enumerators share some abstract shape, but not much else. I wonder if it would be clearer/easier to name them if they were not associated by inheritance at all, but rather just side-by-side implementations? That is, just `TypeEnumerator` and `TypeEnumeratorWithoutSubstituted`. The `Enumerate` overload for `TemplateArgument` would not be necessary for the latter.

Reply:
I might not fully understand your naming suggestion.

Reply:
Sorry, I see how that was unclear.

I mean that it could be nicer to separate these two AST visitors entirely, and not use inheritance to reuse code between them. I don't think there's any strict is-a relationship between them, and the reuse seems limited to a few lines.

But I could be underestimating the wins from sharing code, maybe it's worth it.